pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1376,6 +1384,10 @@ func (r *ReconcileClusterDeployment) syncDeletedClusterDeployment(cd *hivev1.Clu
 func (r *ReconcileClusterDeployment) addClusterDeploymentFinalizer(cd *hivev1.ClusterDeployment) error {
 	cd = cd.DeepCopy()
 	controllerutils.AddFinalizer(cd, hivev1.FinalizerDeprovision)
+	if cd.Spec.ClusterPoolRef != nil && cd.Spec.ClusterPoolRef.CustomizationRef != nil && !controllerutils.HasFinalizer(cd, hivev1.FinalizerCustomizationRelease) {
+		controllerutils.AddFinalizer(cd, hivev1.FinalizerCustomizationRelease)
+	}
```

Comment:
Mm, we can't just fold this into `addClusterDeploymentFinalizer` like this. Consider a scenario where we upgrade to/past this commit when a ClusterDeployment CR already existed and had the deprovision finalizer. Due to the check on L563 we would never get here and thus never add the customization finalizer.

ISTR commenting in this area before and suggesting consolidating in some way, but I must not have said it right :)

What we should do is have a single function that takes the finalizer string as a parameter and calls `AddFinalizer` and `Update`. Then call that function from two independent code paths -- one that looks like L563-571 for the deprovision finalizer; and another that looks similar (minus the metric reporting) for the customization finalizer.

If you wanted to get clever, you could make the function subsume the condition and logging; but then it needs to return a bool indicating whether we did something or not so the caller can return iff we did and continue otherwise.